File: 000001-000200/000017.电话号码的字母组合/17.js
/**
 * @param {string} digits
 * @return {string[]}
 */
// 方法一：用时 64ms
// 执行用时超过了 85%
// 消耗内存超过了 83%
// 难度: 中等
var letterCombinations = function (digits) {
  if (!digits.length) {
    return []
  }
  let letters = ['', '', 'abc', 'def', 'ghi', 'jkl', 'mno', 'pqrs', 'tuv', 'wxyz']
  let ret = []
  for (let i = 0; i < digits.length; i++) {
    let index = digits[i]
    if (!ret.length) {
      ret = letters[index].split('')
      continue
    }
    let tmp = []
    for (let k = 0; k < ret.length; k++) {
      for (let j = 0; j < letters[index].length; j++) {
        tmp.push(ret[k] + letters[index][j])
      }
    }
    ret = tmp
  }
  return ret
};

console.log(letterCombinations('23'))
// console.log(letterCombinations('2'))
// console.log(letterCombinations(''))

File: 000001-000200/000017.电话号码的字母组合/17.go
package main

import (
	"fmt"
	"strconv"
	"strings"
)

// 方法一：用时 0ms
// 执行用时超过了 100%
// 消耗内存超过了 22%
// 难度: 中等
func letterCombinations(digits string) []string {
	if len(digits) == 0 {
		return []string{}
	}
	letters := []string{"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"}
	i0, _ := strconv.Atoi(string(digits[0]))
	ret := strings.Split(letters[i0], "")
	for i := 1; i < len(digits); i++ {
		ii, _ := strconv.Atoi(string(digits[i]))
		tmp := []string{}
		for k := 0; k < len(ret); k++ {
			for j := 0; j < len(letters[ii]); j++ {
				tmp = append(tmp, ret[k] + string(letters[ii][j]))
			}
		}
		ret = tmp
	}
	return ret
}

func main ()  {
	fmt.Println(letterCombinations("23"))
}
